### ISSUE TITLE
Feature request: Database reflection / introspection

### ISSUE THREAD

**Issue by cognifloyd** [ponyorm/pony#440]
PonyORM looks fantastic. It would be even more useful with some kind of (rudimentary) introspection tool to help in developing entities from an existing database.

I have a crufty old database I need to work with, and pony's generator syntax should make reasoning about it much simpler. Generating a rough set of entities would be extremely helpful, and then I can further tune them to smooth over the weirder bits of the database.
If some kind of auto-introspection were made available, I would want to declare a set of entity models, and then have pony (or complementary lib) generate all but those models (with the option to exclude a named list of tables).

[abetkin/pony-inspect](https://github.com/abetkin/pony-inspect) looks interesting, though it is two years old, and doesn't support finding composite primary keys or many-to-many relations (readme is out of date, code suggests it supports mysql and sqlite in addition to postgres). 

Previous requests:
_quotes are responses from @kozlovsky_

https://stackoverflow.com/questions/24675920/how-to-reflect-database-objects-using-pony-orm
- > We can add support of database reflection in the future. (July 2014)
- > reflection probably will not be implemented soon, because legacy database layouts can be really weird, and it is not easy to write algorithm which auto-generate nice entities from weird schemas. (Sep 2014)
- > At this moment Pony still does not support reflexion from the database. Right now we are working on migrations, and after that we can add reflexion capabilities. (Oct 2016)

https://mailman-mail5.webfaction.com/pipermail/ponyorm-list/2017-October/000292.html
- > At this moment, Pony does not have an official way to introspect a
database and reconstruct entities from the database. Typically, a database
does not contain enough information to fully restore entity classes anyway. (Oct 2017)

#244 seems to be sort of similar, though I don't understand what the solution `adict` has to do with introspecting the database.

**Comment by jgirardet** [comment]
You can already use existing databases creating your own entity and specifying table name

**Comment by cognifloyd** [comment]
Yes. And I will begin with that. Tooling to help me generate those pony models would be great (I'm working with a very large MySQL db). The next step would be to introspect the database and compare with my pony models to see if there are any tables or columns I missed when mapping the db to pony.

**Comment by cognifloyd** [comment]
Another interesting resource: peewee's introspector
https://github.com/coleifer/peewee/blob/master/playhouse/reflection.py

I could probably subclass Introspector and override generate_models with something that will generate pony models instead.

**Comment by jgirardet** [comment]
you also have the private` MyEntity._attrs_` or `My.Entity._columns_` which could be usefulll at introspection.

**Comment by cognifloyd** [comment]
To build the database entity models from an existing database, I played with both approaches:
- peewee's [inspector](https://github.com/coleifer/peewee/blob/master/playhouse/reflection.py) (or similar) could be used to dynamically build the model (python object) so that it can be used right away.
- [abetkin/pony-inspect](https://github.com/abetkin/pony-inspect) builds the entity models by stringing together a python file (like a template), so that the generated file must be imported before the models can be used.

Building the models programmatically (like peewee) does not seem possible at this point as there is no way to apply `PrimaryKey`, `composite_index`, or `composite_key` to an object you are building. These assume that they are being called (they are imperative instead of declarative) within the scope of the Entity Model Class (ie they look for sibling attributes in the `f_locals` of the python frame that is assumed to be the entity class).

Here's the relevant functions that make building the entities programmaticly difficult:

https://github.com/ponyorm/pony/blob/8587f885c8058c291272cf55ee13752eaa85f971/pony/orm/core.py#L2623-L2664

I suppose whatever introspector could manipulate `model._indexes_` directly and set the bools on the relevant attrs, duplicating the actual logic to add the keys/indexes here:

https://github.com/ponyorm/pony/blob/8587f885c8058c291272cf55ee13752eaa85f971/pony/orm/core.py#L2627-L2628
https://github.com/ponyorm/pony/blob/8587f885c8058c291272cf55ee13752eaa85f971/pony/orm/core.py#L2659-L2664

So, pony-inspect is more likely to produce something usable based on pony's architecture here, but the implementation in that project is incomplete and so does not work out of the box (at least for MySQL. I haven't tested sqlite which the readme says should work).

**Comment by luckydonald** [comment]
Currently there is an effort for [migration tooling](https://blog.ponyorm.org/2019/03/19/pony-orm-migrations/) going on. That should enable programatic creation of  databases. With that it could be possible to use that peewee inspector to generate the models.
